### ISSUE TITLE
Need too much time

### ISSUE THREAD

**Issue by Ironbrotherstyle** [ChiWeiHsiao/DeepVO_PyTorch#6]
I am a noob and had trouble training the net beause of limited resouce. Could some one offer me a trained model to test the result? that would be very nice, THX.

**Comment by alexart13** [comment]
@Ironbrotherstyle you can find the links to the trained model in the readme

**Comment by Ironbrotherstyle** [comment]
Thank you

**Comment by Ironbrotherstyle** [comment]
![route_04_gradient](https://user-images.githubusercontent.com/28485294/52952709-67386f80-33c0-11e9-894d-7dfdb7bf90ca.png)
![route_09_gradient](https://user-images.githubusercontent.com/28485294/52952719-715a6e00-33c0-11e9-8517-17495a00ebe4.png)
I used the trained model offered by the author and got results on video 04, 05, 07, 09, 10. Only to find that the results on these videos were terrible except for on video04, as shown in the figures. The MSEs also showd the same trend. I am puzzled about this and don not know the reason.  @alexart13 @ChiWeiHsiao @daiyk

**Comment by alexart13** [comment]
@Ironbrotherstyle , did you use the model mentioned in https://github.com/ChiWeiHsiao/DeepVO-pytorch#download-trained-model section?

**Comment by Ironbrotherstyle** [comment]
> @Ironbrotherstyle , did you use the model mentioned in https://github.com/ChiWeiHsiao/DeepVO-pytorch#download-trained-model section?

yes, the file names 't000102050809_v04060710_im184x608_s5x7_b8_rnn1000_optAdagrad_lr0.0005.model.train' , I just noticed that you had uploaed four .py files. Could you please tell me the improvement by these changes like adding to 15 instead of 6 GT poses.

**Comment by alexart13** [comment]
@Ironbrotherstyle you can find description of the changes at https://github.com/ChiWeiHsiao/DeepVO-pytorch/issues/4 Regarding GT file, I also included rotation matrix 'as is' from KITTI data. It is used in ImageSequenceDataset class to align every sequence.

**Comment by alexart13** [comment]
@Ironbrotherstyle I have just tried to reproduce the experiment using the code from this repo. It works as expected. I only updated the following parameters in params.py.
self.img_means =  (0.19007764876619865, 0.15170388157131237, 0.10659445665650864) 		self.img_stds =  (0.2610784009469139, 0.25729316928935814, 0.25163823815039915) 

![route_07_gradient](https://user-images.githubusercontent.com/32840403/53002374-9e993180-343d-11e9-9cba-837562e55fc2.png)

**Comment by Ironbrotherstyle** [comment]
> @Ironbrotherstyle I have just tried to reproduce the experiment using the code from this repo. It works as expected. I only updated the following parameters in params.py.
> self.img_means = (0.19007764876619865, 0.15170388157131237, 0.10659445665650864) self.img_stds = (0.2610784009469139, 0.25729316928935814, 0.25163823815039915)
> 
> ![route_07_gradient](https://user-images.githubusercontent.com/32840403/53002374-9e993180-343d-11e9-9cba-837562e55fc2.png)

I just used the trained model by yourself  in 'https://drive.google.com/uc?id=1l0s3rYWgN8bL0Fyofee8IhN-0knxJF22&export=download' and got results like these pictures below. And i will change the' self.img_means' as you mentioned to see the results later. Thank you!
![route_04_gradient](https://user-images.githubusercontent.com/28485294/53002734-5c391a00-3468-11e9-9f06-ceb4884fd3c4.png)
![route_05_gradient](https://user-images.githubusercontent.com/28485294/53002737-5e9b7400-3468-11e9-98e4-283c024fa8d4.png)

**Comment by alexart13** [comment]
@Ironbrotherstyle setting img_means and img_stds should help. I overlooked to update them in the repo. I created a pull request https://github.com/ChiWeiHsiao/DeepVO-pytorch/pull/9 to fix that.

**Comment by Ironbrotherstyle** [comment]
> @Ironbrotherstyle setting img_means and img_stds should help. I overlooked to update them in the repo. I created a pull request #9 to fix that.

I am really sorry to trouble you so much, but I still have a few questions about the methods in function 'ImageSequenceDataset'. The way you got relative t is like $t^{`}=R_0^{T}(t_1-t_0)$, which is different from what I thought should be by using $t_1=R^{`}t_{0}+t^{`}$, here $t_0, t_1$ are the transformation parts in world coordinate. Can you give me some instructions or links, that would help me a lot.

**Comment by alexart13** [comment]
@Ironbrotherstyle the key idea of the transformation is to let the model 'thinks' that every sequence starts at 0 point and has 0 Euler angles initially. So we need to rotate **the whole** sequence in the opposite direction of the first point rotation matrix R.

**Comment by Ironbrotherstyle** [comment]
I  roughly what you mean. But when I write down the formula, I can’t explain the R0 transpose here when I want to transform t to world. 



    


            On 02/21/2019 17:13, Alexander wrote: @Ironbrotherstyle the key idea of the transformation is to let the model 'thinks' that every sequence starts at 0 point and has 0 Euler angles initially. So we need to rotate the whole sequence in the opposite direction of the first point rotation matrix R.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.
{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/ChiWeiHsiao/DeepVO-pytorch","title":"ChiWeiHsiao/DeepVO-pytorch","subtitle":"GitHub repository","main_image_url":"https://github.githubassets.com/images/email/message_cards/header.png","avatar_image_url":"https://github.githubassets.com/images/email/message_cards/avatar.png","action":{"name":"Open in GitHub","url":"https://github.com/ChiWeiHsiao/DeepVO-pytorch"}},"updates":{"snippets":[{"icon":"PERSON","message":"@alexart13 in #6: @Ironbrotherstyle the key idea of the transformation is to let the model 'thinks' that every sequence starts at 0 point and has 0 Euler angles initially. So we need to rotate **the whole** sequence in the opposite direction of the first point rotation matrix R."}],"action":{"name":"View Issue","url":"https://github.com/ChiWeiHsiao/DeepVO-pytorch/issues/6#issuecomment-465921304"}}}
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/ChiWeiHsiao/DeepVO-pytorch/issues/6#issuecomment-465921304",
"url": "https://github.com/ChiWeiHsiao/DeepVO-pytorch/issues/6#issuecomment-465921304",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by Ironbrotherstyle** [comment]
> @Ironbrotherstyle the key idea of the transformation is to let the model 'thinks' that every sequence starts at 0 point and has 0 Euler angles initially. So we need to rotate **the whole** sequence in the opposite direction of the first point rotation matrix R.

Thank you so much. I`ve solved the problem because I reversed the left multiplication and the 
right multiplication. Now I got your formula!

**Comment by alexart13** [comment]
@Ironbrotherstyle ,
As you know, in KITTI dataset every track starts with camera looking along Z axis. X axis is oriented to the right. Imagine that the car made a 90 degrees right turn and moved along X axis as the result. You take a point from this track after the turn is made. Theta angles will look like 0,-90,0 equal to rotation matrix R. Imagine you want to make this point on the track the new starting point for the rest of the track. You need to make the following calculation:
1) t_n_new = t_n - t_0, where t_0 are the coordinates of the new starting point
2) rotate the tail of the track 90 degrees backward to align the starting point camera view along Z axis. For that you need transposed R matrix. t_n_new_rotated = R_transposed_0 * t_n_new
3) In general case we should also recalculate R and theta for the rest of the track points. But we don't have to do that in our case, because we operate with angle deltas between the two consecutive points on the track:
`# get relative pose w.r.t. previous frame
        groundtruth_sequence[2:] = groundtruth_sequence[2:] - groundtruth_sequence[1:-1]`    

I hope this helps.

**Comment by Ironbrotherstyle** [comment]
> @Ironbrotherstyle ,
> As you know, in KITTI dataset every track starts with camera looking along Z axis. X axis is oriented to the right. Imagine that the car made a 90 degrees right turn and moved along X axis as the result. You take a point from this track after the turn is made. Theta angles will look like 0,-90,0 equal to rotation matrix R. Imagine you want to make this point on the track the new starting point for the rest of the track. You need to make the following calculation:
> 
> 1. t_n_new = t_n - t_0, where t_0 are the coordinates of the new starting point
> 2. rotate the tail of the track 90 degrees backward to align the starting point camera view along Z axis. For that you need transposed R matrix. t_n_new_rotated = R_transposed_0 * t_n_new
> 3. In general case we should also recalculate R and theta for the rest of the track points. But we don't have to do that in our case, because we operate with angle deltas between the two consecutive points on the track:
>    `# get relative pose w.r.t. previous frame groundtruth_sequence[2:] = groundtruth_sequence[2:] - groundtruth_sequence[1:-1]`
> 
> I hope this helps.

Very nice of you. I understand that, thank you so much.

**Comment by Ironbrotherstyle** [comment]
> @Ironbrotherstyle ,
> As you know, in KITTI dataset every track starts with camera looking along Z axis. X axis is oriented to the right. Imagine that the car made a 90 degrees right turn and moved along X axis as the result. You take a point from this track after the turn is made. Theta angles will look like 0,-90,0 equal to rotation matrix R. Imagine you want to make this point on the track the new starting point for the rest of the track. You need to make the following calculation:
> 
> 1. t_n_new = t_n - t_0, where t_0 are the coordinates of the new starting point
> 2. rotate the tail of the track 90 degrees backward to align the starting point camera view along Z axis. For that you need transposed R matrix. t_n_new_rotated = R_transposed_0 * t_n_new
> 3. In general case we should also recalculate R and theta for the rest of the track points. But we don't have to do that in our case, because we operate with angle deltas between the two consecutive points on the track:
>    `# get relative pose w.r.t. previous frame groundtruth_sequence[2:] = groundtruth_sequence[2:] - groundtruth_sequence[1:-1]`
> 
> I hope this helps.

Hi, I tried to retrain the model in the repository, but the validation loss was very abnormal. Have you ever encountered this problem? By the way, can you give me some tricks on training or
![screenshot from 2019-02-23 21-35-46](https://user-images.githubusercontent.com/28485294/53294289-c530e100-381e-11e9-8c95-b19afeca63e9.png)
 setting parameters? Thank you so much.

**Comment by alexart13** [comment]
@Ironbrotherstyle I used the parameters as they were specified in the repository. 
It is difficult to say what might be wrong in your case without knowing all the details.

**Comment by chenmy17** [comment]
Hi，I had a question when i visualize my results, i got this curve. Why ur result curves are so smooth?
Did i ignore something?
![seq_09](https://user-images.githubusercontent.com/33723872/53621494-aa50d900-3c31-11e9-91b9-71267af541b0.png)

**Comment by Ironbrotherstyle** [comment]
I only used the visualize.py in the repo. Make sure that your code is the
same as that in the repo.

chenmingyue <notifications@github.com> 于2019年3月1日周五 下午2:52写道：

**Comment by Ironbrotherstyle** [comment]
> @Ironbrotherstyle setting img_means and img_stds should help. I overlooked to update them in the repo. I created a pull request #9 to fix that.

Hi, I could not understand the calculation of img_means and img_std, your values are not the same as the results got from function "calculate_rgd_mean_std" in preprocess.py which all have negative img_means. So could you please tell me the way to get your img_means and img_stds ?

**Comment by alexart13** [comment]
@Ironbrotherstyle, I dont' remember exactly. I might have used a subset of the dataset to calculate the values to save time. That is why I think you get different results.

**Comment by Ironbrotherstyle** [comment]
> @Ironbrotherstyle, I dont' remember exactly. I might have used a subset of the dataset to calculate the values to save time. That is why I think you get different results.
The img_means and img_std seem to be important, and the performance by using your values are much better than these negative values. Is your program the same as the repository on calculating means and stds and did you use the "minus_point_5=True" ? Thank you for your reply. Hope you have a nice day.

**Comment by alexart13** [comment]
@Ironbrotherstyle The performance is better because the model was trained using these particular img_means and img_std. If you retrain the model using your values, you should get good results too.

I used the method to calculate the values as it is in the repository. minus_point_5 was set to True. But as I mentioned, I might have removed some tracks from the calculation.

**Comment by Ironbrotherstyle** [comment]
> @Ironbrotherstyle The performance is better because the model was trained using these particular img_means and img_std. If you retrain the model using your values, you should get good results too.
> 
> I used the method to calculate the values as it is in the repository. minus_point_5 was set to True. But as I mentioned, I might have removed some tracks from the calculation.

Very nice of you, thanks.

**Comment by ujgygy656565** [comment]
@Ironbrotherstyle 
My results are the same as before. Do you know where the problem is now?
